File: packages/libsql/ios/Plugin/Classes/Options/ExecuteBatchOptions.swift
import Foundation
import Capacitor

@objc public class ExecuteBatchOptions: NSObject {
    let connectionId: String
    let statement: [String]
    let values: [[Any]]?

    init(_ call: CAPPluginCall) throws {
        guard let connectionId = call.getString("connectionId") else {
            throw LibsqlError.connectionIdMissing
        }
        guard let statementArray = call.getArray("statement") else {
            throw LibsqlError.statementMissing
        }
        
        var statements: [String] = []
        for item in statementArray.capacitor.replacingNullValues() {
            guard let statement = item as? String else {
                throw LibsqlError.invalidStatement
            }
            statements.append(statement)
        }
        
        self.connectionId = connectionId
        self.statement = statements
        
        if let valuesArray = call.getArray("values") {
            var allValues: [[Any]] = []
            for item in valuesArray.capacitor.replacingNullValues() {
                guard let subArray = item as? [Any] else {
                    throw LibsqlError.invalidValues
                }
                allValues.append(subArray)
            }
            self.values = allValues
        } else {
            self.values = nil
        }
    }
}

File: packages/libsql/android/src/main/java/io/capawesome/capacitorjs/plugins/libsql/classes/options/ExecuteBatchOptions.java
package io.capawesome.capacitorjs.plugins.libsql.classes.options;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.JSArray;
import com.getcapacitor.PluginCall;
import java.util.ArrayList;
import java.util.List;

public class ExecuteBatchOptions {

    @NonNull
    private final String connectionId;

    @NonNull
    private final List<String> statement;

    @Nullable
    private final List<List<Object>> values;

    public ExecuteBatchOptions(@NonNull PluginCall call) throws Exception {
        this.connectionId = getConnectionIdFromCall(call);
        this.statement = getStatementFromCall(call);
        this.values = getValuesFromCall(call);
    }

    @NonNull
    public String getConnectionId() {
        return connectionId;
    }

    @NonNull
    public List<String> getStatement() {
        return statement;
    }

    @Nullable
    public List<List<Object>> getValues() {
        return values;
    }

    @NonNull
    private static String getConnectionIdFromCall(@NonNull PluginCall call) throws Exception {
        String connectionId = call.getString("connectionId");
        if (connectionId == null) {
            throw new Exception("connectionId must be provided.");
        }
        return connectionId;
    }

    @NonNull
    private static List<String> getStatementFromCall(@NonNull PluginCall call) throws Exception {
        JSArray statementArray = call.getArray("statement");
        if (statementArray == null) {
            throw new Exception("statement must be provided.");
        }

        List<String> statements = new ArrayList<>();
        for (int i = 0; i < statementArray.length(); i++) {
            Object item = statementArray.get(i);
            if (item instanceof String) {
                statements.add((String) item);
            } else {
                throw new Exception("All statements must be strings.");
            }
        }
        return statements;
    }

    @Nullable
    private static List<List<Object>> getValuesFromCall(@NonNull PluginCall call) throws Exception {
        JSArray valuesArray = call.getArray("values");
        if (valuesArray == null) {
            return null;
        }

        List<List<Object>> allValues = new ArrayList<>();
        for (int i = 0; i < valuesArray.length(); i++) {
            Object item = valuesArray.get(i);
            if (item instanceof JSArray) {
                JSArray subArray = (JSArray) item;
                List<Object> subValues = new ArrayList<>();
                for (int j = 0; j < subArray.length(); j++) {
                    subValues.add(subArray.get(j));
                }
                allValues.add(subValues);
            } else {
                throw new Exception("All values must be arrays.");
            }
        }
        return allValues;
    }
}
